### PR TITLE
Refactored dioxus components

### LINKED ISSUE #1: Usage of Dioxus components
hey! Very interesting project, I hope you continue with it.

I see you are currently separating the UI in a sort type of Components but not exactly, it should be improved :D

You are currently doing it like this:

```rs
fn app() -> Element {
   rsx!(
       button { "hey" }
       {whatever(123)}
   )
}

fn whatever(value: usize) -> Element {
   rsx!(
       p { "whatever -> {value}" }
   )
}
```

But it should be like this:

```rs
fn app() -> Element {
   rsx!(
       button { "hey" }
       Whatever {
          value: 123
       }
   )
}

#[component]
fn Whatever(value: usize) -> Element {
   rsx!(
       p { "whatever -> {value}" }
   )
}
```

:v: 😄

### PR DESCRIPTION
**Pull request by albbus-stack**
Implements dioxus components properly, following the suggestion of issue #1

### CODE REVIEW

File: src/ui.rs [hunk 1]
```diff
@@ -1,37 +1,41 @@
...
 }
 
-pub fn button(
-    text: Element,
-    onclick: impl FnMut(Event<MouseData>) + 'static,
-    class: &str,
-    variant: Variant,
-) -> Element {
+#[component]
+pub fn TextButton(props: ButtonProps) -> Element {
```
Comment (marc2332) [inline src/ui.rs]: When using `#[component]` you can inline the props as if they were normal parameters without the need of a separate struct for the props:

```rs
#[component]
pub fn TextButton(text: String, onclick: EventHandler<MouseEvent>, class: String, variant: Variant) -> Element {
```

Or you can keep the `ButtonProps` struct and remove `#[component]`, both work the same way :)

File: src/ui.rs [hunk 2]
```diff
@@ -42,17 +46,24 @@ pub fn button(
                         "bg-gray-100 shadow-gray-200/10 hover:shadow-gray-200/20 text-gray-900"
                     }
                 },
-                class,
+                &props.class,
             ]
                 .join(" "),
-            onclick: onclick,
-            {text}
+            onclick: move |evt| onclick.clone().call(evt),
```
Comment (marc2332) [inline src/ui.rs]: Not sure whether this clone is needed:

```suggestion
            onclick: move |evt| onclick.call(evt),
```

File: src/ui.rs [hunk 3]
```diff
@@ -42,17 +46,24 @@ pub fn button(
...
+pub struct IconProps {
+    pub name: String,
+    pub class: String,
+}
+
+#[component]
+pub fn Icon(props: IconProps) -> Element {
     rsx! {
-        span { class: ["material-icons", class].join(" "), {name} }
+        span { class: ["material-icons", &props.class].join(" "), {props.name.clone()} }
```
Comment (marc2332) [inline src/ui.rs]:
```suggestion
        span { class: ["material-icons", &props.class].join(" "), {props.name} }
```

File: src/ui.rs [hunk 4]
```diff
@@ -42,17 +46,24 @@ pub fn button(
...
 
-pub fn icon(name: &str, class: &str) -> Element {
+#[derive(PartialEq, Props, Clone)]
+pub struct IconProps {
+    pub name: String,
+    pub class: String,
+}
+
+#[component]
+pub fn Icon(props: IconProps) -> Element {
```
Comment (marc2332) [inline src/ui.rs]: Same for the props

File: src/pages/about.rs [hunk 5]
```diff
@@ -1,4 +1,5 @@
-use crate::{get_head, ui, Context, Page};
+use crate::ui::*;
+use crate::{get_head, Context, Page};
 use dioxus::prelude::*;
 
 pub fn page(mut context: Context) -> Element {
```
Comment (marc2332) [inline src/pages/about.rs]: You should also convert the pages to actual components

**Review by marc2332 — CHANGES_REQUESTED** [review]

**Comment by marc2332** [comment]
As a bonus, you should probably use dioxus-router for the routing as well, but it's just a nice-to-have

**Comment by albbus-stack** [comment]
> As a bonus, you should probably use dioxus-router for the routing as well, but it's just a nice-to-have

I'll also look into that, thanks for the interest! You rock 🤘

**Pull request merged**